lib/shared-method.js

```diff
@@ -303,30 +304,37 @@ function coerceAccepts(uarg, desc) {
     });
   }
 
+  var argIsNullOrUndefined = uarg === null || uarg === undefined;
   var actualType = SharedMethod.getType(uarg);
 
   // convert values to the correct type
   // TODO(bajtos) Move conversions to HttpContext (and friends)
   // SharedMethod should only check that argument values match argument types.
   var conversionNeeded = targetType !== 'any' &&
-    actualType !== 'undefined' &&
+    !argIsNullOrUndefined &&
     actualType !== targetType;
 
   if (conversionNeeded) {
-    // JSON.parse can throw, so catch this error.
-    try {
-      uarg = convertValueToTargetType(name, uarg, targetType);
-      actualType = SharedMethod.getType(uarg);
-    } catch (e) {
-      var message = util.format('invalid value for argument \'%s\' of type ' +
-        '\'%s\': %s. Received type was %s. Error: %s',
-        name, targetType, uarg, typeof uarg, e.message);
-      throw new badArgumentError(message);
+    if (Dynamic.canConvert(targetType)) {
+      // Convert input so it always matches the method definition.
+      // The user can also define new dynamic types.
+      uarg = new Dynamic(uarg).to(targetType);
+    } else {
+      // JSON.parse can throw, so catch this error.
+      try {
+        uarg = convertValueToTargetType(name, uarg, targetType);
+      } catch (e) {
+        var message = util.format('Invalid value for argument \'%s\' of type ' +
+          '\'%s\'. Received type was %s. Error: %s',
+          name, targetType, typeof uarg, e.message);
+        throw new badArgumentError(message);
```

Comment:
This is classic reflection, and production APIs simply shouldn't be echoing request data back. Yes, there are ways to mitigate it, and yes, you shouldn't be displaying it in the UI, but there are always edge cases (like [MIME sniffing](https://adblockplus.org/blog/the-hazards-of-mime-sniffing)) that can make this dangerous. With this reflection hole open, in certain cases, XSS would be possible via a specially-crashed API link.